### ISSUE TITLE
[Grouped product] Can´t add simple products to cart if one other is out of stock

### ISSUE THREAD

**Issue by hamplr** [magento/magento2#22304]
### Preconditions (*)
1. magento 2.2-develop
2. php 7.1

### Steps to reproduce (*)
- enable the option allow to show products without stock 
- grouped product with more sub items
- at least one sub items have to be without stock (quantity=0)
- navigate to the PDS of the grouped product
- add at least on item of any available sub item
- if you add the item to the cart the message "Please specify the quantity of product(s)." apears

### Expected result (*)
1. the simple products are added to the cart

### Actual result (*)
1. the message "Please specify the quantity of product(s)." appears
2. it´s not possible to add a simple product, if one other simple product within the grouped product is out of stock

**Comment by hamplr** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by larsroettig** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by larsroettig** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @hamplr. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @larsroettig. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @larsroettig. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @hamplr, here is your [Magento instance](https://i-22304-2-2-develop.instances.magento-community.engineering).
Admin access: [https://i-22304-2-2-develop.instances.magento-community.engineering/admin](https://i-22304-2-2-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by nickpiro** [comment]
Having the same issue in 2.2.8

**Comment by larsroettig** [comment]
Hi @nickpiro,
we have a fix as Patch for 2.2.8 currently it is not possible to fix this without breaking the [Backward compatiblty](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/). Maybee we will find way but I must ask the Team,

```diff
Index: project/src/vendor/magento/module-grouped-product/Model/Product/Type/Grouped.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- vendor/magento/module-grouped-product/Model/Product/Type/Grouped.php
+++ vendor/magento/module-grouped-product/Model/Product/Type/Grouped.php
@@ -338,9 +338,6 @@
         }
         foreach ($associatedProducts as $subProduct) {
             if (!isset($productsInfo[$subProduct->getId()])) {
-                if ($isStrictProcessMode && !$subProduct->getQty()) {
-                    return __('Please specify the quantity of product(s).')->render();
-                }
                 $productsInfo[$subProduct->getId()] = $subProduct->isSalable() ? (float)$subProduct->getQty() : 0;
             }
         }
```

**Comment by celorodovalho** [comment]
I have the same problem here in 2.2.8
Any solution?
This is a blocking issue!

@larsroettig there is a commit for this patch?

**Comment by hamplr** [comment]
No, there´s currently no commit for this patch, but we have the patch in use in our project to solve the problem.

**Comment by mallikaverma** [comment]
@magento-engcom-team  give me 2.3-develop instance

**Comment by mallikaverma** [comment]
I am working on it

**Comment by mallikaverma** [comment]
I am working on this #event-imagine2019

**Comment by mallikaverma** [comment]
#imagine2019

**Comment by joekersh** [comment]
The fix from @larsroettig works for me, i created a preference for that class and then overrode the getProductInfo method:

In your module di.xml:

`<preference for="Magento\GroupedProduct\Model\Product\Type\Grouped" type="Vendor\Module\Model\Grouped"/>`

New class Vendor\Module\Model\Grouped:


	protected function getProductInfo(\Magento\Framework\DataObject $buyRequest, $product, $isStrictProcessMode)
    {
        $productsInfo = $buyRequest->getSuperGroup() ?: [];
        $associatedProducts = $this->getAssociatedProducts($product);

        if (!is_array($productsInfo)) {
            return __('Please specify the quantity of product(s).')->render();
        }
        foreach ($associatedProducts as $subProduct) {
            if (!isset($productsInfo[$subProduct->getId()])) {
            	
                // if ($isStrictProcessMode && !$subProduct->getQty()) {
                //     return __('Please specify the quantity of product(s).')->render();
                // }
                
                $productsInfo[$subProduct->getId()] = $subProduct->isSalable() ? (float)$subProduct->getQty() : 0;
            }
        }

        return $productsInfo;
    }

`

**Comment by larsroettig** [comment]
@joekersh preference is not the best way to solve this issue patching is a better way yesterday @mallikaverma  creating a pr this look backwards compatible.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @dmytro-ch 
 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-99841` were created

**Issue Available**: @dmytro-ch, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by reense** [comment]
I can confirm that @larsroettig 's solution is working. Thanks !

**Comment by VladimirZaets** [comment]
Hi @hamplr. Thank you for your report. 
The issue has been fixed in magento/magento2#24955 by @gihovani in 2.4-develop branch
Related commit(s):
 - [f27f74950a0fb961834bd4909015718e228a8ddf](https://github.com/magento/magento2/commit/f27f74950a0fb961834bd4909015718e228a8ddf)
 - [1e2ee963657c1c0a7d4c692acb0b16c97dd971e0](https://github.com/magento/magento2/commit/1e2ee963657c1c0a7d4c692acb0b16c97dd971e0)
 - [c6d6aaba1cc94f3d13101326ebbf011ec430ac37](https://github.com/magento/magento2/commit/c6d6aaba1cc94f3d13101326ebbf011ec430ac37)
 - [907f5820dad7400d595af5ed086cb5b7d3855f50](https://github.com/magento/magento2/commit/907f5820dad7400d595af5ed086cb5b7d3855f50)

The fix will be available with the upcoming 2.4.0 release.
